core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -354,26 +376,54 @@ protected RelNode decorrelate(RelNode root) {
   private HepPlanner createPlanner(HepProgram program) {
     // Create a planner with a hook to update the mapping tables when a
     // node is copied when it is registered.
-    return new HepPlanner(
-        program,
-        context,
-        true,
-        createCopyHook(),
-        RelOptCostImpl.FACTORY);
+    HepPlanner planner =
+        new HepPlanner(
+            program,
+            context,
+            true,
+            createCopyHook(),
+            RelOptCostImpl.FACTORY);
+    planner.setDecorrelator(this);
+    return planner;
   }
 
+  /**
+   * Tries to remove the {@link org.apache.calcite.rel.core.Correlate} instances of a query
```

Comment:
I think "removes some instances of Correlate from a query plan" is a bit more precise.